File: baseunits/modules/KuManga.pas
unit KuManga;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, Math, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, synautil, URIParser;

implementation

const
  dirurl = '/backend/ajax/searchengine.php';
  dirperpage = '15';
  dirpostdata = 'contentType=manga&retrieveCategories=true&retrieveAuthors=true&perPage=' +
    dirperpage + '&page=';

function GetDirectoryPageNumber(const MangaInfo: TMangaInformation;
  var Page: Integer; const WorkPtr: Integer; const Module: TModuleContainer): Integer;
var
  s: String;
begin
  Result := NET_PROBLEM;
  Page := 1;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  if MangaInfo.FHTTP.POST(Module.RootURL + dirurl, dirpostdata + '1') then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        s := XPathString('json(*).totalContents');
        Page := ceil(StrToIntDef(s, 1) / StrToInt(dirperpage));
      finally
        Free;
      end;
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  c, i: Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  if MangaInfo.FHTTP.POST(Module.RootURL + dirurl, dirpostdata + IncStr(AURL)) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        c := XPath('json(*).contents()').Count;
        for i := 1 to c do
        begin
          ANames.Add(XPathString('json(*).contents(' + IntToStr(i) + ').name'));
          ALinks.Add(XPathString('json(*).contents(' + IntToStr(i) + ')/concat("/manga/",id,"/",slug)'));
        end;
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  lastPage, i, totalChapters, perPage: Integer;
  s: String;
  tmp: TStringList;
  uri: TURI;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo do
  begin
    url := FillHost(Module.RootURL, AURL);
    if MangaInfo.FHTTP.GET(url) then begin
      Result := NO_ERROR;
      tmp := TStringList.Create;
      lastPage := 0;
      with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
        try
          coverLink := XPathString('//div[@class="row"]//img/@src');
          if coverLink <> '' then coverLink := MaybeFillHost(Module.RootURL, coverLink);
          if title = '' then title := XPathString('//h1');
          summary := XPathString('//*[@class="infom"]/div/div[1]/p');
          authors := XPathString('//*[@class="infom"]/div/div[2]//p[contains(.,"Autor") and not(contains(.,"No Disponible"))]/substring-after(normalize-space(.),": ")');
          artists := XPathString('//*[@class="infom"]/div/div[2]//p[contains(.,"Artist") and not(contains(.,"No Disponible"))]/substring-after(normalize-space(.),": ")');
          status := MangaInfoStatusIfPos(XPathString('//*[@class="infom"]/div/div[2]//p[contains(.,"Estado")]'), 'Activo', 'Finalizado');
          genres := XPathString('//*[@class="panel-footer" and contains(.,"Géneros")]/string-join(.//a,", ")');
          XPathHREFAll('//table[contains(@class, "table")]//h4[@class="title"]/a', chapterLinks, chapterName);

          s := XPathString('//script[contains(., "php_pagination")]');
          if s <> '' then begin
            s := GetBetween('php_pagination(', ');', s);
            tmp.Delimiter := ',';
            tmp.DelimitedText := s;
            totalChapters := StrToIntDef(tmp[4], 0);
            perPage := StrToIntDef(tmp[5], 1);
            lastPage := Math.ceil(float(totalChapters) / float(perPage));
          end;
        finally
          tmp.Free;
          Free;
        end;

      for i := 2 to lastPage do begin
        uri := ParseURI(url);
        uri.Path := uri.Path + 'p/' + IntToStr(i);
        s := EncodeURI(uri);
        if MangaInfo.FHTTP.GET(s) then begin
           with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
           try
             XPathHREFAll('//table[contains(@class, "table")]//h4[@class="title"]/a', chapterLinks, chapterName);
           finally
             Free;
           end;
        end;
      end;

      InvertStrings([chapterLinks, chapterName]);
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  pages: Integer;
  pageFormat, pageFormat2: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container, DownloadThread.FHTTP do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(MaybeFillHost(Module.RootURL, AURL.Replace('/c/','/leer/'))) then
    begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do try
        pageFormat := XPathString('//script[contains(.,"konekomangareader")]/substring-before(substring-after(substring-after(.,"setup"),","),");")');
        if pageFormat <> '' then begin
          pageFormat := ReplaceString(pageFormat, ':remote', ':''remote''');
          ParseHTML(pageFormat);
          pages := StrToIntDef(XPathString('json(*)//pages'), 0);
          pageFormat := XPathString('substring-before(json(*)//pageFormat,"{pnumber}")');
          pageFormat2 := XPathString('substring-after(json(*)//pageFormat,"{pnumber}")');
          if pageFormat <> '' then
            for pages := 1 to pages do
              PageLinks.Add(pageFormat + IntToStr(pages) + pageFormat2);
        end;
      finally
        Free;
      end;
    end;
  end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'KuManga';
    RootURL := 'http://www.kumanga.com';
    Category := 'Spanish';
    OnGetDirectoryPageNumber := @GetDirectoryPageNumber;
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/KuManga.lua
----------------------------------------------------------------------------------------------------
-- Scripting Parameters
----------------------------------------------------------------------------------------------------

local LuaDebug   = require 'LuaDebugging'
-- LuaDebugging  = true   --> Override the global LuaDebugging variable by uncommenting this line.
-- LuaStatistics = true   --> Override the global LuaStatistics variable by uncommenting this line.


----------------------------------------------------------------------------------------------------
-- Local Constants
----------------------------------------------------------------------------------------------------

local DirectoryPagination = '/backend/ajax/searchengine.php'
local DirectoryParameters = 'contentType=manga&retrieveCategories=true&retrieveAuthors=true&perPage=18&page='


----------------------------------------------------------------------------------------------------
-- Event Functions
----------------------------------------------------------------------------------------------------

-- Get info and chapter list for current manga.
function GetInfo()
  local c, x = nil
  local u = MaybeFillHost(module.RootURL, url)
  
  --[[Debug]] LuaDebug.WriteLogWithHeader('GetInfo', 'url ->  ' .. u)
  if not http.Get(u) then return net_problem end
  
  x = TXQuery.Create(http.Document)
  mangainfo.url       = x.XPathString('//meta[@property="og:url"]/@content')
  mangainfo.Title     = x.XPathString('//meta[@property="og:title"]/@content')
  mangainfo.CoverLink = x.XPathString('//meta[@property="og:image"]/@content')
  mangainfo.Status    = MangaInfoStatusIfPos(x.XPathString('//*[@class="infom"]/div/div[2]//p[contains(.,"Estado")]'), 'Activo', 'Finalizado')
  mangainfo.Authors   = x.XPathString('//*[@class="infom"]/div/div[2]//p[contains(.,"Autor") and not(contains(.,"No Disponible"))]/substring-after(normalize-space(.),": ")')
  mangainfo.Artists   = x.XPathString('//*[@class="infom"]/div/div[2]//p[contains(.,"Artist") and not(contains(.,"No Disponible"))]/substring-after(normalize-space(.),": ")')
  mangainfo.Genres    = x.XPathString('//*[@class="panel-footer" and contains(.,"Géneros")]/string-join(.//a,", ")')
  mangainfo.Summary   = x.XPathString('//meta[@property="og:description"]/@content')
  
  x.XPathHREFAll('//table[contains(@class, "table")]//h4[@class="title"]/a', mangainfo.ChapterLinks, mangainfo.ChapterNames)
  c = GetPageChapterListPageCount(x.XPathString('//script[contains(., "php_pagination")]'))
  
  if c > 1 then
    for i = 2, c do
      if http.Get(mangainfo.url .. '/p/' .. i) then
        x = TXQuery.Create(http.Document)
        x.XPathHREFAll('//table[contains(@class, "table")]//h4[@class="title"]/a', mangainfo.ChapterLinks, mangainfo.ChapterNames)
      end
    end
  end
  
  InvertStrings(mangainfo.ChapterLinks, mangainfo.ChapterNames)
  
  --[[Debug]] LuaDebug.PrintMangaInfo()
  --[[Debug]] LuaDebug.WriteStatistics('Chapters', mangainfo.ChapterLinks.Count .. '  (' .. mangainfo.Title .. ')')
  
  return no_error
end


-- Get the page count of the manga list of the current website.
function GetDirectoryPageNumber()
  local u = module.RootURL .. DirectoryPagination
  
  --[[Debug]] LuaDebug.WriteLogWithHeader('GetDirectoryPageNumber', 'url ->  ' .. u .. '   (Params: [' .. DirectoryParameters .. '])')
  if not http.Post(u, DirectoryParameters .. '1') then return net_problem end
  
  page = math.ceil(tonumber(TXQuery.Create(http.Document).XPathString('json(*).totalContents')) / 18)
  
  --[[Debug]] LuaDebug.WriteStatistics('DirectoryPages', page)
  
  return no_error
end


-- Get links and names from the manga list of the current website.
function GetNameAndLink()
  local c, x = nil
  local u = module.RootURL .. DirectoryPagination
  
  --[[Debug]] LuaDebug.WriteLogWithHeader('GetNameAndLink', 'url ->  ' .. u .. '   (Params: [' .. DirectoryParameters .. '])')
  if not http.Post(u, DirectoryParameters .. IncStr(url)) then return net_problem end
  
  x = TXQuery.Create(http.Document)
  c = x.XPathCount('json(*).contents()')
  
  for i = 1, c do
    names.Add(x.XPathString('json(*).contents(' .. i .. ').name'))
    links.Add(x.XPathString('json(*).contents(' .. i .. ')/concat("/manga/",id,"/",slug)'))
  end
  
  --[[Debug]] LuaDebug.PrintMangaDirectoryEntries(DirectoryParameters .. IncStr(url))
  
  return no_error
end


-- Get the page count for the current chapter.
function GetPageNumber()
  local c, p, x = nil
  local u = MaybeFillHost(module.RootURL, url:gsub('/c/', '/leer/'))
  
  --[[Debug]] LuaDebug.WriteLogWithHeader('GetPageNumber', 'url ->  ' .. u)
  if not http.Get(u) then return net_problem end
  
  x = TXQuery.Create(http.Document)
  c = tonumber(x.XPathString('//script[contains(., "konekomangareader")]'):match('\'pages\':(.-),'))
  p = x.XPathString('//script[contains(., "konekomangareader")]'):match('\'pageFormat\':\'(.-)\',')
  
  for i = 1, c do
    task.PageLinks.Add(p:gsub('{pnumber}', tostring(i)))
  end
  
  --[[Debug]] LuaDebug.PrintChapterPageLinks()
  --[[Debug]] LuaDebug.WriteStatistics('ChapterPages', task.PageLinks.Count .. '  (' .. u .. ')')
  
  return no_error
end


----------------------------------------------------------------------------------------------------
-- Helper Functions
----------------------------------------------------------------------------------------------------

-- Get count of chapter list pages.
function GetPageChapterListPageCount(s)
  local count = 1
  
  s = GetBetween('php_pagination(', ');', s)
  count = math.ceil(tonumber(s:match('.-,.-,.-,.-,(.-),.-,.-')) / tonumber(s:match('.-,.-,.-,.-,.-,(.-),.-')))
  
  return count
end


----------------------------------------------------------------------------------------------------
-- Module Initialization
----------------------------------------------------------------------------------------------------

function Init()
  AddWebsiteModule('KuManga', 'http://www.kumanga.com', 'Spanish')
end

function AddWebsiteModule(name, url, category)
  local m = NewModule()
  m.Website                  = name
  m.RootURL                  = url
  m.Category                 = category
  m.OnGetInfo                = 'GetInfo'
  m.OnGetNameAndLink         = 'GetNameAndLink'
  m.OnGetPageNumber          = 'GetPageNumber'
  m.OnGetDirectoryPageNumber = 'GetDirectoryPageNumber'
  return m
end